### ISSUE TITLE
LVM subsystem optimized calls & parsing in `lvm.go` by removing stateful view

### ISSUE THREAD

**Issue by jakobmoellerdev** [topolvm/topolvm#800]
**What should the feature do:**

We want to optimize the way that TopoLVM interacts with LVM to make it less resource-intensive for Create/Update/Delete operations and not result in CPU Spikes that are caused by tons of updates in its LVM view.

Currently TopoLVM relies on the so called [LVM State](https://github.com/topolvm/topolvm/blob/main/lvmd/command/lvm_json.go#L162) for almost all of its interactions with the LVM subsystem.

Almost all of the LVM commands trigger the Update of its world-view of LVM in-memory, which is done in https://github.com/topolvm/topolvm/blob/main/lvmd/command/lvm.go#L69-L83.

Not only does every command trigger the entire JSON view of the LVM world, but it also iterates over all LVS for every volume group to update them.

We can optimize the behavior by leveraging LVMs existing promise to not have side effects with its commands. This means that extending a thin-pool will never have any effect on any other thin-pool. Also updating a volume group will never touch any other volume other than the specified one.

That means that we can change all calls from "get everything in LVM" to "only get the updated logical volumes / volume groups / thin pool data for the objects we touched".

**What is use case behind this feature:**

The changes result in way less CPU/Memory overhead per call as we would no longer need to refresh all of our LVM understanding on every call. It will avoid CPU / Memory Spikes on PVC creation / deletion and will enhance scalability of the system both in Data Centre and Edge Use cases.

It will lead to:

1. No longer have to ask LVM to get all of LVs, VGs at once
2. No longer have a possibly huge JSON to parse for every command
3. No longer have to spend memory on maintaing a "World View" of LVM

Basically we can get rid of `Update` and `filter_lv` entirely and focus on just updating the necessary parts on demand. Instead we can focus on a stateless model. Here is a reference of how this can look like: https://github.com/openshift/lvm-operator/blob/main/internal/controllers/vgmanager/lvm/lvm.go based on LVM Operator.

If there is interest in this kind of change, I would like to volunteer for the change and testing adjustments required. 
If there are doubts about the change, feel free to let me know here.

**Comment by llamerada-jp** [comment]
Thanks for your proposal.

I confirmed that the `Update` and `FindVolumeGroup` methods fetch every volume group information. As you point out, this is too much, so I believe optimization would be effective.

The `vgService.send` method is executed periodically to get data for metrics. Multiple volume information is fetched using lvm command only once for each timing. Executing lvm commands individually for each volume group can result in performance degradation. Is it correct that this process is not included in this proposal because it does not match Create/Update/Delete?

Additional points that could be improved:
It is a waste to not free the `vgs` used by [this process](https://github.com/topolvm/topolvm/blob/d40d797e046b949fee716b6833e4595b6bee97c4/pkg/lvmd/cmd/root.go#L61-L81), so simply moving the process to another function may reduce memory usage if golang compiler is not optimized enough.

**Comment by jakobmoellerdev** [comment]
I believe that for the metrics service, we should try to have only one single call to LVM so having the LVMState command for this once would be sufficient in my opinion. I agree that we should focus on optimizing the calls made for individual LVs/VGs only.

I will prepare a draft next week.

**Comment by jakobmoellerdev** [comment]
I have a PR upcoming and plan to put it up for review. I wanted to ask if we have a reference load for performance analysis that I can use for profile comparison? I did some internal testing for verification and can share these results otherwise but they are made with lvm operator together for ease of setup.

**Comment by llamerada-jp** [comment]
We didn't have a reference load until for performance analysis. With the change of PR, is expected to speed up the create PV. I think the following script can measure the effect using e2e test env. I think it is difficult to measure for the deletion method, but since it is a similar process, I believe it has a similar effect.

```sh
kubectl delete sts perf
kubectl delete pvc -l app=perf

kubectl apply -f - <<'EOF'
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: perf
spec:
  selector:
    matchLabels:
      app: perf
  replicas: 100
  serviceName: perf
  template:
    metadata:
      labels:
        app: perf
    spec:
      containers:
      - name: pause
        image: registry.k8s.io/pause
        volumeMounts:
        - name: perf
          mountPath: /mnt
  volumeClaimTemplates:
  - metadata:
      name: perf
    spec:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: "topolvm-provisioner"
      resources:
        requests:
          storage: 32Mi
EOF

DURATIONS=""
PVCS=$(kubectl get pvc -l app=perf -o jsonpath='{.items[*].metadata.name}')
for PVC in $PVCS; do
  PVC_CREATION_TIME=$(kubectl get pvc $PVC -o jsonpath='{.metadata.creationTimestamp}')
  PV=$(kubectl get pvc $PVC -o jsonpath='{.spec.volumeName}')
  PV_CREATION_TIME=$(kubectl get pv $PV -o jsonpath='{.metadata.creationTimestamp}')
  DURATION=$(expr $(date -d $PV_CREATION_TIME +%s) - $(date -d $PVC_CREATION_TIME +%s))
  DURATIONS="$DURATIONS $DURATION"
done

SORTED=$(echo $DURATIONS | tr ' ' '\n' | sort -n)
# drop head 5 items and tail 5 items
CUTED=$(echo $SORTED | cut -d ' ' -f 6-95)

MIN=$(echo $CUTED | cut -d ' ' -f 1)
MAX=$(echo $CUTED | cut -d ' ' -f 90)
SUM=0
for DURATION in $CUTED; do
  SUM=$(expr $SUM + $DURATION)
done
AVG=`echo "scale=3; $SUM / 90" | bc`

echo "$MIN $MAX $AVG"
```
